### PR TITLE
Processing acceleration of full row ROI

### PR DESCRIPTION
**Pull request by 0x72D0**
PR for #87 

the function work with any function except the projection profile function

### CODE REVIEW

File: src/image_function_simd.cpp [hunk 1]
```diff
@@ -1678,6 +1678,50 @@ if ( simdType == neon_function ) { \
...
+    }
+
+    template <typename Type, typename TImage, typename... Args>
+    bool widthCondition(const Type & width, TImage image, Args... args)
+    {
+        return image.rowSize() == width && widthCondition(width, args...);
+    }
+
+    template <typename Type>
+    void OptimiseROI( Type & height, Type & width )
```
Comment (ihhub) [inline src/image_function_simd.cpp]: With my proposed code changes we would have only 1 `OptimiseROI` function:
```cpp
    template <typename... Args>
    void OptimiseROI(uint32_t & width uint32_t & height, Args... args)
    {
        if( IsFullImageRow(width, image, args...) && ( width < (std::numeric_limits<uint32_t>::max() / height) ) ) {
            width = width * height;
            height = 1u;
        }
    }
```

**Comment by ihhub** [comment]
@0x72D0 the best way to verify such performance enhancement is to modify performance tests. For **SSE or NEON** (because your processor is slower) we can create images of 255 pixels width and 10000 pixels height and test before our changes and after.

File: src/image_function_simd.cpp [hunk 2]
```diff
@@ -1678,6 +1678,50 @@ if ( simdType == neon_function ) { \
...
 #endif
 
+
+    template <typename Type, typename TImage, typename... Args>
+    bool widthCondition(const Type & width, TImage image)
+    {
+        return image.rowSize() == width;
+    }
+
+    template <typename Type>
```
[the remark this answers is not in the document]
Comment (0x72D0) [reply src/image_function_simd.cpp]: which function?

**Comment by 0x72D0** [comment]
I made the changes, but for some reason there's a segmentation to the sse version of the code

**Comment by ihhub** [comment]
All tests failed too. I'm looking into this issue.

**Comment by ihhub** [comment]
I found the root of the problem:
For example in `void AbsoluteDifference` function we modify **width** and **height** before calculating `const uint8_t * outYEnd = outY + height * rowSizeOut;`. Once we set `height` as 1 we never hit a condition that our pointer would be equal to **outYEnd** leading to infinite loop and segmentation fault.
The solution is to set `OptimiseROI` after calculating **outYEnd** variable.

**Comment by ihhub** [comment]
Additionally, `height` and `width` were swapped as input parameters.

**Comment by ihhub** [comment]
On top of this we forgot to change `rowSize` in all functions.

**Comment by 0x72D0** [comment]
we don't need to change the rowSize, the only error was the swapped input parameter. If we put the OptimizationRoi after the parameterValidation, it's good

**Comment by ihhub** [comment]
@0x72D0 could you please make performance tests in the way as I described previously?

**Comment by ihhub** [comment]
I did testings myself and results are very promising. This is before optimisation:
```
[44/57] image_function_sse::AbsoluteDifference (255x255)... 1.027+/-0.08774 ms
[45/57] image_function_sse::Accumulate (255x255)... 0.163+/-0.03057 ms
[46/57] image_function_sse::BitwiseAnd (255x255)... 1.026+/-0.087 ms
[47/57] image_function_sse::BitwiseOr (255x255)... 0.8922+/-0.2718 ms
[48/57] image_function_sse::BitwiseXor (255x255)... 0.8365+/-0.05608 ms
[49/57] image_function_sse::Invert (255x255)... 0.1205+/-0.02562 ms
[50/57] image_function_sse::Maximum (255x255)... 0.8505+/-0.08669 ms
[51/57] image_function_sse::Minimum (255x255)... 0.8422+/-0.06522 ms
[52/57] image_function_sse::ProjectionProfile (255x255)... 0.1009+/-0.0236 ms
[53/57] image_function_sse::RgbToBgr (255x255)... 1.928+/-0.1455 ms
[54/57] image_function_sse::Subtract (255x255)... 0.8582+/-0.05428 ms
[55/57] image_function_sse::Sum (255x255)... 0.07514+/-0.01545 ms
[56/57] image_function_sse::Threshold (255x255)... 0.1582+/-0.02871 ms
[57/57] image_function_sse::ThresholdDouble (255x255)... 0.1853+/-0.02919 ms
```
And this is after optimisation:
```
[44/57] image_function_sse::AbsoluteDifference (255x255)... 0.5985+/-0.06085 ms
[45/57] image_function_sse::Accumulate (255x255)... 0.1504+/-0.05575 ms
[46/57] image_function_sse::BitwiseAnd (255x255)... 0.6108+/-0.1129 ms
[47/57] image_function_sse::BitwiseOr (255x255)... 0.5494+/-0.1298 ms
[48/57] image_function_sse::BitwiseXor (255x255)... 0.5249+/-0.02941 ms
[49/57] image_function_sse::Invert (255x255)... 0.1032+/-0.02394 ms
[50/57] image_function_sse::Maximum (255x255)... 0.5231+/-0.04049 ms
[51/57] image_function_sse::Minimum (255x255)... 0.5254+/-0.02906 ms
[52/57] image_function_sse::ProjectionProfile (255x255)... 0.1015+/-0.01453 ms
[53/57] image_function_sse::RgbToBgr (255x255)... 1.917+/-0.1715 ms
[54/57] image_function_sse::Subtract (255x255)... 0.5284+/-0.04898 ms
[55/57] image_function_sse::Sum (255x255)... 0.04522+/-0.005681 ms
[56/57] image_function_sse::Threshold (255x255)... 0.139+/-0.0247 ms
[57/57] image_function_sse::ThresholdDouble (255x255)... 0.1474+/-0.02943 ms
```

**Comment by ihhub** [comment]
@0x72D0 thank you very much for your work on the project!

I have an idea how to extend it to normal CPU functions as well to gain an extra performance. I'll create an issue later for it.

**Pull request merged**